src/core/include/openvino/core/graph_util.hpp

```diff
@@ -219,11 +220,20 @@ void replace_nodes(const std::shared_ptr<Model>& f,
                        parameter_replacement_map,
                    const std::unordered_map<std::shared_ptr<Node>, std::shared_ptr<Node>>& body_replacement_map);
 
+namespace {
+struct EdgeHash {
+    inline size_t operator()(const std::tuple<Node*, Node*, size_t>& v) const {
+        return (size_t(std::get<0>(v)) << 2) + (size_t(std::get<1>(v)) << 1) + std::get<2>(v);
+    }
+};
+}  // namespace
+
 /// Topological sort of nodes needed to compute root_nodes
 template <typename T>
 std::vector<std::shared_ptr<Node>> topological_sort(T root_nodes) {
     std::stack<Node*, std::vector<Node*>> nodes_to_do;
     std::unordered_set<Node*> nodes_done;
+    std::unordered_set<std::tuple<Node* /*from*/, Node* /*to*/, size_t /*edge_dx*/>, EdgeHash> visited_edges;
```

Comment:
why we can't store just <Node*, size_t> ? Do we actually need the additional Node* ?

Reply:
I believe we can improve the solution if we will use 
unordered_map<Node*, std::bitset<num_inputs>> nodes_done;
instead of the current
std::unordered_set<Node*> nodes_done;

Right now you store 2 maps (nodes_done, visited_edges). 
nodes_done map memory consumption ~ sizeof(Node*) * sizeof(hash) * node_cnt
visited_edges memory consumption ~ sizeof(tuple<Node*, Node*, size_t>) * sizeof(hash) * edge_cnt
I believe visited_edges map is not necessary here.

If we will use 
unordered_map<Node*, std::bitset<num_inputs>> 
memory consumption will be ~ sizeof(Node*) * sizeof(hash) * sizeof(bitset) * node_cnt

As for time, I believe the access to already preallocated memory in bitset (size of bitset we know because of we know count of inputs for each node) and bit operations will be faster when inserting each edge to unordered_set and custom hash calculation.

Please check it locally.

Reply:
Done - I applied same idea, but without std::bitset in order to not indexing edges